experimental/parser/parse_decl.go

```diff
@@ -0,0 +1,695 @@
+		// parse a range instead. Rewinding is necessary because otherwise we get
+		// into an annoying situation where if we have e.g. reserved foo to bar;
+		// we have already consumed reserved foo, but we want to push foo
+		// through the expression machinery to get foo to bar as a single
+		// expression.
+		c.Rewind(mark)
+		return parseRange(p, c).AsAny()
+	}
+
+	return parseDef(p, c, kw, ty, path).AsAny()
+}
+
+// parseBody parses a ({}-delimited) body of declarations.
+func parseBody(p *parser, braces token.Token, in taxa.Subject) ast.DeclBody {
+	body := p.NewDeclBody(braces)
+
+	// Drain the contents of the body into it. Remember,
+	// parseDecl must always make progress if there is more to
+	// parse.
+	c := braces.Children()
+	for !c.Done() {
+		if next := parseDecl(p, c, in); !next.Nil() {
+			body.Append(next)
+		}
+	}
+
+	return body
+}
+
+// parseRange parses a reserved/extensions range.
+func parseRange(p *parser, c *token.Cursor) ast.DeclRange {
+	// Consume the keyword token.
+	kw := c.Pop()
+
+	in := taxa.Extensions
+	if kw.Text() == "reserved" {
+		in = taxa.Reserved
+	}
+
+	// Consume expressions until we hit a semicolon or the [ of a compact
+	// options. Note that this means that we do not parse reserved [1, 2, 3];
+	// "correctly".
+	var (
+		// badExpr keeps track of whether we exited the loop due to a parse
+		// error or because we hit ; or [ or EOF.
+		badExpr bool
+		exprs   []exprComma
+	)
+
+	// Parse expressions until we hit a semicolon or the [ of compact options.
+	elems := commas(p, c, true, func(c *token.Cursor) (ast.ExprAny, bool) {
+		next := c.Peek().Text()
+		if next == ";" || next == "[" {
+			return ast.ExprAny{}, false
+		}
+
+		expr := parseExpr(p, c, in.In())
+		badExpr = badExpr || expr.Nil()
+
+		return expr, !expr.Nil()
+	})
+	elems(func(expr ast.ExprAny, comma token.Token) bool {
+		exprs = append(exprs, exprComma{expr, comma})
+		return true
+	})
+
+	var options ast.CompactOptions
+	if next := c.Peek(); next.Text() == "[" && !next.IsLeaf() {
+		options = parseOptions(p, c.Pop(), in)
+	}
+
+	// Parse a semicolon, if possible.
+	semi, err := p.Punct(c, ";", in.After())
+	if err != nil && !badExpr {
+		p.Error(err)
+	}
+
+	r := p.NewDeclRange(ast.DeclRangeArgs{
+		Keyword:   kw,
+		Options:   options,
+		Semicolon: semi,
+	})
+	for _, e := range exprs {
+		r.AppendComma(e.expr, e.comma)
+	}
+
+	return r
+}
+
+// parseDef parses a generic definition.
+func parseDef(p *parser, c *token.Cursor, kw token.Token, ty ast.TypeAny, path ast.Path) ast.DeclDef {
+	args := ast.DeclDefArgs{
+		Type: ty,
+		Name: path,
+	}
+
+	var (
+		inputs, outputs token.Token
+		braces          token.Token
+
+		outputTy ast.TypeAny // Used only for diagnostics.
+	)
+
+	// Try to parse the various "followers". We try to parse as many as
+	// possible: if we have `foo = 5 = 6`, we want to parse the second = 6,
+	// diagnose it, and throw it away.
+	var mark token.CursorMark
+followers:
```

Comment:
I left the label here, because I think it makes the code simple. I don't really understand why you have a problem with break-to-label but not break-without-label, especially because Go made the mistake of keeping break for switch, which, combined with lints that want to turn if-elses into switches, means that it is not obvious where a break is breaking to a lot of the time.

I also don't think that goto is bad, personally, but I'm just used to C++ where you have to use goto to break out of outer loops, since C++ doesn't have break-to-label.

Reply:
We have different aesthetic tastes. But also different experiences: I've spent 9+ years on teams that strictly disallowed labeled break/goto for the reasons I mentioned above: they are often a smell and the logic can be more easily understood by reorganizing code or factoring things out into separate functions/methods. While I see what you're saying about break-from-switch causing issues when you want to break from a loop, and a labeled break being a convenient solution, I'm still in a habit of just never using labeled break. I don't mind constructions to work around that, even if they are sometimes awkward. I am also pretty certain that no other codebase at Buf would allow this, so avoiding it is an organizational norm, too.